bids/layout/layout.py

```diff
@@ -342,20 +334,48 @@ def do_begin(conn):
 
         self.session = sa.orm.sessionmaker(bind=engine)()
 
-    def _init_db(self, database_file=None, reset_database=False):
+    def _init_db(self, database_dir=None, reset_database=False):
+        if database_dir is not None:
+            database_file = os.path.join(database_dir, 'layout_index.sqlilte')
+            os.makedirs(database_dir, exist_ok=True)
+            database_sidecar = os.path.join(database_dir, 'layout_args.json')
+        else:
+            database_file = None
+            database_sidecar = None
+
         # Reset database if needed and return whether or not it was reset
         # determining if the database needs resetting must be done prior
-        # prior to setting the session (which creates the empty database file)
-        reset_database = (reset_database or  # Manual Request
-                          not database_file or  # In memory transient database
-                          not os.path.exists(database_file))  # New file based database created
+        # to setting the session (which creates the empty database file)
+        reset_database = (
+            reset_database or  # Manual Request
+            not database_dir or  # In memory transient db
+            not os.path.exists(database_file)  # New file based db created
+        )
+
+        # Prepare instance arguments for serialization
+        instance_args = self.__dict__.copy()
+        for k in ['ignore', 'force_index']:
+            instance_args[k] = [str(a) for a in instance_args[k]]
+        instance_args.pop('sources', None)
 
         self._set_session(database_file)
 
-        if reset_database:
+        if not reset_database:
+            saved_args = json.load(open(database_sidecar))
+            for k, v in instance_args.items():
+                if saved_args[k] != v:
+                    raise ValueError(
+                        "Initaliation arguments do not match for database dir:"
+                        f" {database_dir}"
+                        )
+        else:
+            if database_sidecar is not None:
```

Comment:
Should probably move the JSON writing to after engine creation, just in case the former fails.